File: Python/check6.py

import sys
sys.stdin = open('../input.txt')

# MAXCITY = 100
# MAXCOUNT = 2000 + 1
INF = 999999
BIAS = 24*60
city_name = {}
trains = [[],[]] # trains, rtrains
from_to = [[],[],[],[]] # fr_h, fr_t, to_h, to_t
ncity = 0
nconn = 0

def m2hm(m, hm):
    hm[0] = int((m / 60))
    hm[1] = m - (60 * hm[0])

# 駅名 -> 駅番号
def city_id(name):
    global INF, ncity, nconn

    # keyが存在すればvalueを返す
    if name in city_name:
        return city_name[name]

    # 要素を追加
    city_name[name] = ncity
    ncity += 1
    return ncity - 1

# trains[0]の用意
def parse_connection(buf):
    global INF, ncity, nconn, start, end

    dpt = [0,0]; arv = [0,0]
    dpt[0],dpt[1] = map(int, buf[1].split(":"))
    arv[0],arv[1] = map(int, buf[3].split(":"))
    if (dpt[0]*60 + dpt[1] < start) or (arv[0]*60 + arv[1] > end):
        return

#     print(dpt[0]," ",dpt[1]," ",arv[0]," ",arv[1])

    # リスト内に辞書を追加
    trains[0].append({})
    trains[0][nconn]["from"] = city_id(buf[0])
    trains[0][nconn]["dpt"] = dpt[0]*60 + dpt[1]
    trains[0][nconn]["to"] = city_id(buf[2])
    trains[0][nconn]["arv"] = arv[0]*60 + arv[1]
    trains[0][nconn]["fare"] = int(buf[4])
    nconn += 1

# trains[1]の用意，ソート
def prepare_data():
    global INF, ncity, nconn, BIAS

    for i in range(0,nconn):
        trains[1].append({})
        trains[1][i]["from"] = trains[0][i]["to"]
        trains[1][i]["to"] = trains[0][i]["from"]
        trains[1][i]["dpt"] = BIAS - trains[0][i]["arv"]
        trains[1][i]["arv"] = BIAS - trains[0][i]["dpt"]
        trains[1][i]["fare"] = trains[0][i]["fare"]

    # "arv"をキーにしてソート
    trains[0].sort(key=lambda x:x["arv"])
    trains[1].sort(key=lambda x:x["arv"])

# 列車配列tvの中で，dpttime以前に駅stに到着する列車を探す
# p: 探索を開始する列車到着事象の番号
def change(tr, p, st, dptTime):
    global INF, ncity, nconn

    while p >= 0:
        if (tr[p]["to"] == st) and (tr[p]["arv"] <= dptTime):
            break
        p -= 1
    return p


# 列車配列tvを参照して，駅orgを起点として，
# fr_h または fr_t を二次元配列vに作成する。
def make_tbl(tbl, org, tr):
    global INF, ncity, nconn

    # tblの初期化
    for i in range(0,ncity):
        tbl.append([])
        for j in range(0,nconn+1):
            tbl[i].append([INF,0,0])
#             tbl[i][0].append(INF)
    tbl[org][0][0] = 0

    # 列車の数だけ更新
    for ti in range(0,nconn):
        # 前の状態を引き継ぐ
        for i in range(0,ncity):
            for j in range(0,3):
                tbl[i][ti+1][j] = tbl[i][ti][j]

        # 列車配列tvの中で，tv[ti].dpt以前に駅tv[ti].fromに到着する列車を探す
        # つまり，列車tv[ti]に乗り継げる列車を探す。
        # ti-1: 探索を開始する列車到着事象の番号
        a = change(tr, ti-1, tr[ti]["from"], tr[ti]["dpt"])

#         #安い方を引き継ぐ
        if tbl[tr[ti]["to"]][ti][0] > tr[ti]["fare"] + tbl[tr[ti]["from"]][a+1][0]:
            tbl[tr[ti]["to"]][ti+1][0] = tr[ti]["fare"] + tbl[tr[ti]["from"]][a+1][0];
            tbl[tr[ti]["to"]][ti+1][1] = tr[ti]["dpt"];
            tbl[tr[ti]["to"]][ti+1][2] = tr[ti]["arv"];
#         else :
#             tbl[tr[ti]["to"]][ti+1][0] = tbl[tr[ti]["to"]][ti][0]

def calc_cost(city, minc):
    global INF, ncity, nconn, BIAS, meetTime

    a = 0
    d = nconn - 1
    minc[0] = INF

    while True:
        # 面会時間
        stay = (BIAS - trains[1][d]["arv"]) - trains[0][a]["arv"]
        if stay < meetTime:
            d -= 1
            if d < 0: break
            continue

        c = 0
        for i in range(0,2): c += from_to[i][city][a+1][0]
        for i in range(0,2): c += from_to[i+2][city][d+1][0]

        if minc[0] > c:
            minc[0] = c
            minc[1] = max(from_to[0][city][a+1][2],from_to[1][city][a+1][2])
            minc[2] = BIAS - max(from_to[2][city][d+1][2],from_to[3][city][d+1][2])

        a += 1
        if a >= nconn: break

# main内のもろもろの処理
def solve(minc):
    global INF, ncity, nconn

    prepare_data();
    for i in range(0,4):
        from_to[i].clear()
        make_tbl(from_to[i], city_id(station[i%2]), trains[int(i/2)])

    minc[0] = INF
    for c in range(0,ncity):
        tmp = [0,0,0]
        calc_cost(c,tmp)
        if minc[0] > tmp[0]:
            for i in range(0,3): minc[i] = tmp[i]
            minc[3] = c

        # チェック4のテスト用
#         for name in city_name:
#             if city_name[name] == c:
#                 print(name,": ",tmp[0])

    if minc[0] >= INF: minc[0] = 0

# 表出力
def print_tbl(num):
    global INF, ncity, nconn, station

    print("From " if num < 2 else "To ",end = "")
    print(station[num%2],":\ni\ttime|",end = "")

    for i in range(0,ncity): print(i,"\t",end = "")
    print("\n----------------+------------------------")
    for i in range(0,nconn+1):
        print(i-1,"\t",end = "")
        if i: print(trains[int(num/2)][i-1]["arv"],end = "")
        print("\t|",end = "")
        for j in range(0,ncity):
            for k in range(0,3):
                print(from_to[num][j][i][k],"\t",end = "")
        print()

    print()

if __name__ == '__main__':

    # 実行引数格納
    argv = sys.argv
    if len(argv) != 6:
        print("argment error")
        sys.exit()

    station = []
    for i in range(0,2): station.append(argv[i+1])
    h,m = map(int, argv[3].split(":"))
    start = h*60 + m
    h,m = map(int, argv[4].split(":"))
    end = h*60 + m
    meetTime = int(argv[5]);

    while True:
        # 変数群のリセット
        ncity = 0
        nconn = 0
        for i in range(0,2): trains[i].clear()
        city_name.clear()

        # 列車数読み込み
        count = int(input())
#         print(count)

        # 終了条件
        if count == 0: break

        # 列車読み込み
        for i in range(0,count):
            buf = input().split()
            parse_connection(buf)

        # main内のもろもろの処理
        minc = [0,0,0,0]
        solve(minc)

        # 表出力
#         for i in range(0,4): print_tbl(i)

        print(minc[0]," ",end = "")
        if minc[0] :
            for name in city_name:
                if city_name[name] == minc[3]:
                    print(name," ",end = "")
            hm = [0,0]
            m2hm(minc[1], hm)
            print("{0:02d}:{1:02d} - ".format(hm[0],hm[1]),end = "")
            m2hm(minc[2], hm)
            print("{0:02d}:{1:02d} ".format(hm[0],hm[1]),end = "")

        print()

File: C/check6.c

#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#define FOR(i,a,b) for(int i=a;i<b;i++)
#define REP(i,n)  FOR(i,0,n)
#define MIN(a,b) (((a)<(b)) ? (a):(b))
#define MAX(a,b) (((a)<(b)) ? (b):(a))

#define MAXCITY 100
#define MAXCOUNT 2000 + 1
#define INFINITE 999999
#define BIAS 24*60 //24h * 60min 1440

char station[2][18];
int start, end, meetTime;

char city_name[MAXCITY][18]; //駅名の表
int ncity; //駅数
int nconn; //列車数

// fr_?, to_? の最低額とその時の出発・到着時間をまとめた配列
int ft[4][MAXCITY][MAXCOUNT][3];

// 列車配列
struct train {
	int from, to;
	int dpt, arv;
	int fare;
} trains[2][MAXCOUNT];//,rtrains[MAXCOUNT];

//駅名 -> 駅番号
int city_id(char *name) {
	FOR(i,0,ncity) {
		if(strcmp(name, &(city_name[i][0])) == 0) return i;
	}
	strcpy(&(city_name[ncity][0]), name);
	return ncity++;
}

// 駅追加
void parse_connection(char *buf) {
	char from[18], to[18];
	int dpt[2], arv[2], fare;

	sscanf(buf,"%s %d:%d %s %d:%d %d",
			from, &(dpt[0]), &(dpt[1]), to, &(arv[0]), &(arv[1]), &fare);

	if(dpt[0]*60 + dpt[1] < start || arv[0]*60 + arv[1] > end) return;

	trains[0][nconn].from = city_id(from);
	trains[0][nconn].to = city_id(to);
	trains[0][nconn].dpt = dpt[0]*60 + dpt[1];
	trains[0][nconn].arv = arv[0]*60 + arv[1];
	trains[0][nconn].fare = fare;
	nconn++;
}

// ソート用
int cmp_arv(const void *t1, const void *t2) {
    return ((struct train*)t1)->arv - ((struct train*)t2)->arv;
}

// 分 -> 何時何分
void m2hm(int min, int hm[]) {
	hm[0] = (int)(min / 60);
	hm[1] = min - 60 * (int)(min / 60);
}

// trains, rtrains配列の準備
void prepare_data(void) {
	REP(i,nconn) {
		trains[1][i].from = trains[0][i].to;
		trains[1][i].to = trains[0][i].from;
		trains[1][i].dpt = BIAS - trains[0][i].arv;
		trains[1][i].arv = BIAS - trains[0][i].dpt;
		trains[1][i].fare = trains[0][i].fare;
	}
	qsort(trains[0], nconn, sizeof(struct train), cmp_arv);
	qsort(trains[1], nconn, sizeof(struct train), cmp_arv);
}

//列車配列tvの中で，dpttime以前に駅stに到着する列車を探す
//p: 探索を開始する列車到着事象の番号
int change(struct train tv[], int p, int st, int dpttime) {
	while(p >= 0) {
		if((tv[p].to == st) && (tv[p].arv <= dpttime)) break;
		p--;
	}
	return p;
}

//列車配列trを参照して，駅orgを起点として，
//fr_h または fr_t を二次元配列ftに作成する。
void make_table(int ft[MAXCITY][MAXCOUNT][3], int org, struct train tr[]) {
	int a;

	REP(i,ncity) ft[i][0][0] = INFINITE;
	ft[org][0][0] = 0;

	REP(ti,nconn) {
		REP(i,ncity) REP(j,3) ft[i][ti+1][j] = ft[i][ti][j];

		//列車配列tvの中で，tv[ti].dpt以前に駅tv[ti].fromに到着する列車を探す
		//つまり，列車tv[ti]に乗り継げる列車を探す。
		//ti-1: 探索を開始する列車到着事象の番号
		a = change(tr, ti-1, tr[ti].from, tr[ti].dpt);

//		v[tv[ti].to][ti+1][0] = MIN(v[tv[ti].to][ti][0], tv[ti].fare+v[tv[ti].from][a+1][0]);
		if(ft[tr[ti].to][ti][0] > tr[ti].fare+ft[tr[ti].from][a+1][0]) {
			ft[tr[ti].to][ti+1][0] = tr[ti].fare+ft[tr[ti].from][a+1][0];
			ft[tr[ti].to][ti+1][1] = tr[ti].dpt;
			ft[tr[ti].to][ti+1][2] = tr[ti].arv;
		}
	}
}

// コスト計算
// min[]はコストとその時の面会開始・終了時間
void calc_cost(int city, int min[]) {
	int a = 0, d = nconn-1;
	int stay;
	int c;

	min[0] = INFINITE;
	while(1) {
		stay = (BIAS - trains[1][d].arv) - trains[0][a].arv;
		if(stay < meetTime) {
			d--;
			if(d < 0) break;
			continue;
		}

		c = 0;
		REP(i,2) c += ft[i][city][a+1][0];
		REP(i,2) c += ft[i+2][city][d+1][0];

		if(c < min[0]) {
			min[0] = c;
			min[1] = MAX(ft[0][city][a+1][2],ft[1][city][a+1][2]);
			min[2] = BIAS - MAX(ft[2][city][d+1][2],ft[3][city][d+1][2]);

//			int hm[2];
//			printf("%s ",city_name[city]);
//			printf("%d ",min[0]);
//			m2hm(min[1], hm);
//			printf("%02d:%02d - ",hm[0],hm[1]);
//			m2hm(min[2], hm);
//			printf("%02d:%02d",hm[0],hm[1]);
//			printf("\n");
		}
		a++;
		if(a >= nconn) break;
	}
}

// コスト最小値とその時の面会時間
void solve(int min[]) {
	int tmp[3];//, min[3];

	//
	prepare_data();
	REP(i,4) make_table(ft[i], city_id(station[i%2]), trains[i/2]);

	min[0] = INFINITE;
	REP(c,ncity) {
		calc_cost(c, tmp);
		if(min[0] > tmp[0]) {
			REP(i,3) min[i] = tmp[i];
			min[3] = c;
		}
//		printf("%s: %d\n",city_name[c],tmp[0]); //チェックポイント４のテスト
	}
	if(min[0] >= INFINITE) min[0] = 0;
}

// 出力用
void print_table(int num) {
	printf(num < 2 ? "From":"To");
	printf(" %s:\ni\ttime|",station[num%2]);
	REP(i,ncity) printf("%d\t",i);
	printf("\n----------------+------------------------\n");
	REP(i,nconn+1) {
		printf("%d\t",i-1);
		if(i) printf("%d",trains[num/2][i-1].arv);
		printf("\t|");
		REP(j,ncity) REP(k,3) printf("%d ",ft[num][j][i][k]);
		printf("\n");
	}
	printf("\n");
}

int main(int argc, char *argv[]) {
	int count, min[4], tmp[2], hm[2];
	char buf[64];

	// 引数を変数に格納
	if(argc != 6) {
		fprintf(stderr, "argc error\n");
		return 1;
	}
	REP(i,2) strcpy(station[i],argv[i+1]);
	sscanf(argv[3],"%d:%d",&tmp[0],&tmp[1]);
	start = tmp[0]*60 + tmp[1];
	sscanf(argv[4],"%d:%d",&tmp[0],&tmp[1]);
	end = tmp[0]*60 + tmp[1];
	meetTime = atoi(argv[5]);

	count = 0;
	while(fgets(buf, sizeof(buf), stdin) != NULL) {
		nconn = 0;
		ncity = 0;

		// 列車読み込み
		sscanf(buf, "%d", &count);
		if(!count) break;
		REP(i,count) {
			if(fgets(buf, sizeof(buf), stdin) == NULL) break;
			parse_connection(buf);
		}

		// コスト最小値とその時の面会時間
		solve(min);
//		REP(i,4) print_table(i);

		// 結果出力
		printf("%d ",min[0]);
		if(min[0]) {
			printf("%s: ",city_name[min[3]]);
			m2hm(min[1], hm);
			printf("%02d:%02d - ",hm[0],hm[1]);
			m2hm(min[2], hm);
			printf("%02d:%02d ",hm[0],hm[1]);
		}
		printf("\n");
	}
	return 0;
}




